tests/integration/security/util/framework.go

```diff
@@ -112,29 +137,79 @@ func SetupApps(ctx resource.Context, i istio.Instance, apps *EchoDeployments, bu
 			},
 		}
 		builder.
-			With(nil, EchoConfig(ASvc, apps.Namespace, false, nil, cluster)).
-			With(nil, EchoConfig(BSvc, apps.Namespace, false, nil, cluster)).
+			With(nil, EchoConfig(ASvc, apps.Namespace1, false, nil, cluster)).
+			With(nil, EchoConfig(BSvc, apps.Namespace1, false, nil, cluster)).
+			With(nil, EchoConfig(CSvc, apps.Namespace1, false, nil, cluster)).
+			With(nil, EchoConfig(DSvc, apps.Namespace1, false, nil, cluster)).
 			With(nil, cfg).
-			With(nil, EchoConfig(NakedSvc, apps.Namespace, false, echo.NewAnnotations().
+			With(nil, EchoConfig(NakedSvc, apps.Namespace1, false, echo.NewAnnotations().
 				SetBool(echo.SidecarInject, false), cluster))
 	}
 	for _, c := range ctx.Clusters().ByNetwork() {
 		// VM specific setup
-		vmCfg := EchoConfig(VMSvc, apps.Namespace, false, nil, c[0])
+		vmCfg := EchoConfig(VMSvc, apps.Namespace1, false, nil, c[0])
 		// for test cases that have `buildVM` off, vm will function like a regular pod
 		vmCfg.DeployAsVM = buildVM
 		builder.With(nil, vmCfg)
-		builder.With(nil, EchoConfig(HeadlessSvc, apps.Namespace, true, nil, c[0]))
-		builder.With(nil, EchoConfig(HeadlessNakedSvc, apps.Namespace, true, echo.NewAnnotations().
+		builder.With(nil, EchoConfig(HeadlessSvc, apps.Namespace1, true, nil, c[0]))
+		builder.With(nil, EchoConfig(HeadlessNakedSvc, apps.Namespace1, true, echo.NewAnnotations().
 			SetBool(echo.SidecarInject, false), c[0]))
 	}
+	for _, cluster := range ctx.Clusters() {
+		builder.
+			With(nil, EchoConfig(BSvc, apps.Namespace2, false, nil, cluster)).
+			With(nil, EchoConfig(CSvc, apps.Namespace2, false, nil, cluster)).
+			With(nil, EchoConfig(XSvc, apps.Namespace2, false, nil, cluster))
+	}
+	portC := 8090
+	ports := []echo.Port{
+		{
+			Name:        "http",
+			Protocol:    protocol.HTTP,
+			ServicePort: 80,
+			// We use a port > 1024 to not require root
+			InstancePort: 8090,
+		},
+	}
+	for _, cluster := range ctx.Clusters() {
+		builder.
+			With(nil, echo.Config{Service: CSvc, Namespace: apps.Namespace3,
+				Subsets: []echo.SubsetConfig{{}},
+				Ports: []echo.Port{
+					{
+						Name:         "http",
+						Protocol:     protocol.HTTP,
+						InstancePort: portC,
+					},
+				},
+				Cluster: cluster}).
+			With(nil, echo.Config{
+				Service:   ESvc,
+				Namespace: apps.Namespace3,
+				Subsets:   []echo.SubsetConfig{{}},
+				Ports:     ports,
+				Cluster:   cluster,
+			}).
+			With(nil, echo.Config{
+				Service:   FSvc,
+				Namespace: apps.Namespace3,
+				Subsets:   []echo.SubsetConfig{{}},
+				Ports:     ports,
+				Cluster:   cluster,
+			})
+	}
 	echos, err := builder.Build()
 	if err != nil {
 		return err
 	}
 	apps.All = echos
 	apps.A = echos.Match(echo.Service(ASvc))
 	apps.B = echos.Match(echo.Service(BSvc))
+	apps.C = echos.Match(echo.Service(CSvc))
```

Comment:
I think its highly confusing that apps.X has multiple namespaces in it. This will lead to many issues I would imagine, where I apply some config for x.namespace and suddenly it doesn't apply to all calls

Reply:
Maybe it is better to filter out the apps here based on namespace, like apps.AinNS1, instead of filtering in the test logic.